ravenframework/Optimizers/parentSelectors/parentSelectors.py

```diff
@@ -62,91 +66,139 @@ def rouletteWheel(population,**kwargs):
+                                                replace=False, engine=None)
+      # Extract relevant information
+      if fitnessProvided:
+        matrixOperationRaw[:, 0] = selectedChromo
+        matrixOperationRaw[:, 1] = np.transpose(fitness[kwargs['objVar'][0]][selectedChromo].values)
+
+      tournamentWinnerIndex = int(matrixOperationRaw[np.argmax(matrixOperationRaw[:, 1]), 0])
+      allSelected.add(tournamentWinnerIndex)
+      selectedParent[i, :] = population.values[tournamentWinnerIndex, :]
+  else: # Multi-objective case
+
+    rankProvided = 'rank' in kwargs
+    crowdDistanceProvided = 'crowdDistance' in kwargs
+
+    if not rankProvided or not crowdDistanceProvided or 'fitness' not in kwargs:
+      # Handle cases where neither fitness nor rank are provided
+      mh.error('parentSelectors',ValueError, 'At least one of "fitness" or "rank" must be provided for multi-objective selection')
+    allSelected = set()
+    for i in range(nParents):
+      if rankProvided and crowdDistanceProvided:
+      # If both rank and crowd distance are provided, use them directly as per NSGA-II
+        matrixOperationRaw = np.zeros((kwargs['kSelection'], 2))
+        selectChromoIndexes = list(set(range(len(population.chromosome))) - allSelected)
+        selectedChromo = randomUtils.randomChoice(selectChromoIndexes, size=kwargs['kSelection'],
+                                                  replace=False, engine=None)
+        # Extract relevant information
+        matrixOperationRaw[:, 0] = np.transpose(kwargs['rank'].data[selectedChromo])
+        matrixOperationRaw[:, 1] = np.transpose(kwargs['crowdDistance'].data[selectedChromo])
+        # Stage 1: Select based on rank and crowding distance
+        minRankIndex = list(np.where(matrixOperationRaw[:, 0] == matrixOperationRaw[:, 0].min())[0])
+        if len(minRankIndex) != 1:
+          # Handle cases where more than one chromosome has the same rank
+          minRankNmaxCDIndex = list(np.where((matrixOperationRaw[minRankIndex, 1] ==
+                                              matrixOperationRaw[minRankIndex, 1].max()) &
+                                             (matrixOperationRaw[minRankIndex, 0] ==
+                                              matrixOperationRaw[minRankIndex, 0].min()))[0])
         else:
-          index = int(matrixOperation[i+1,0])
-      selectedParent[i,:] = pop.values[index,:]
-
+          minRankNmaxCDIndex = minRankIndex
+        # Stage 2: Select the individual with the highest crowding distance within their rank group
+        tournamentWinnerIndex = int(minRankNmaxCDIndex[0])
+      elif rankProvided and not crowdDistanceProvided:
+        # If only rank is provided (without crowd distance), calculate a default crowding distance
+        matrixOperationRaw = np.zeros((kwargs['kSelection'], 1))
+        selectChromoIndexes = list(set(range(len(population))) - allSelected)
```

Comment:
same here for "set, range and len"?

Reply:
Fixed in e72143b9463bf4d3929259611658c40247c7a5b5